Monero: add more legacy verify functions

Boog900 opened the pull request:
Fixes #367 

- Adds support for verifying legacy ring signatures
- Changes blocks major/ miner versions to u8 to [match monero](https://github.com/monero-project/monero/blob/master/src/cryptonote_basic/cryptonote_basic.h#L459-L460)
- Changes more of the varint code to be generic over the integer type
- adds back support for multiple input aggregate MLSAG signatures
- adds verifying for both types of MLSAG signatures 
- adds an `UnreducedScalar` type which has a method for recovering scalars which underwent an incorrect reduction

Boog900 committed:
Add v1 ring sig verifying

Boog900 committed:
allow calculating signature hash for v1 txs

Boog900 committed:
add unreduced scalar type with recovery

I have added this type for borromen sigs, the ee field can be a normal
scalar as in the verify function the ee
field is checked against a reduced scalar mean for it to verify as
correct ee must be reduced

Boog900 committed:
change block major/ minor versions to u8

this matches Monero

I have also changed a couple varint functions to accept the `VarInt`
trait

Boog900 committed:
expose `serialize_hashable` on `Block`

Boog900 committed:
add back MLSAG verifying functions

I still need to revert the commit removing support for >1 input MLSAG FULL

This adds a new rct type to separate Full and simple rct

Boog900 committed:
add back support for multiple inputs for RCT FULL

Boog900 committed:
comment `non_adjacent_form` function

also added `#[allow(clippy::needless_range_loop)]` around a loop as without a re-write satisfying clippy without it will make the function worse.

Boog900 committed:
Improve Mlsag verifying API

Boog900 committed:
fix rebase errors

Boog900 committed:
revert the changes on `reserialize_chain`
plus other misc changes

kayabaNerve commented:
no-std CI failure is legitimate, processor/coordinator/full-stack aren't (they're some Docker connectivity issue). I've re-run them.

kayabaNerve commented on coins/monero/src/serialize.rs:
```diff
@@ -29,8 +29,9 @@ pub(crate) fn write_byte<W: Write>(byte: &u8, w: &mut W) -> io::Result<()> {
-  let mut varint = *varint;
+// This will panic if the VarInt exceeds u64::MAX
+pub(crate) fn write_varint<W: Write, U: sealed::VarInt>(varint: &U, w: &mut W) -> io::Result<()> {
+  let mut varint: u64 = (*varint).try_into().map_err(|_| "varint exceeded u64").unwrap();
```
`varint = u64::try_from(varint).unwrap()` should be cleaner than try_into with a type annotation.

kayabaNerve commented on coins/monero/src/block.rs:
```diff
@@ -79,7 +79,7 @@ impl Block {
   }
 
-  fn serialize_hashable(&self) -> Vec<u8> {
+  pub fn serialize_hashable(&self) -> Vec<u8> {
```
Why add pub here?

Boog900 replied:
This is used to calculate the proof of work hash and IMO it's better to get users to provide their own POW hashing functions than to include one in monero-serai

kayabaNerve commented on coins/monero/src/unreduced_scalar.rs:
```diff
@@ -0,0 +1,171 @@
+use core::cmp::Ordering;
+
+use std_shims::io::{self, *};
+use std_shims::sync::OnceLock;
```
`use std_shims::{sync::OnceLock, io::{self, *}};`

kayabaNerve commented on coins/monero/src/unreduced_scalar.rs:
```diff
@@ -0,0 +1,171 @@
+  }
+}
+
+#[cfg(test)]
```
`coins/monero/src/tests/unreduced_scalar.rs`?

kayabaNerve commented on coins/monero/src/ringct/mlsag.rs:
```diff
@@ -25,47 +34,163 @@ impl Mlsag {
     })
   }
 
-  #[cfg(feature = "experimental")]
```
I'd prefer to keep this as experimental. It may not be knowingly incorrect anymore, yet I'd prefer to discuss it being signed off on as proper in a distinct issue/PR.

kayabaNerve commented on coins/monero/src/ringct/borromean.rs:
```diff
@@ -1,15 +1,11 @@
 
-#[cfg(feature = "experimental")]
 use monero_generators::H_pow_2;
-#[cfg(feature = "experimental")]
```
I'd prefer to keep this experimental, another comment describes why.

kayabaNerve commented on coins/monero/src/ringct/borromean.rs:
```diff
@@ -20,54 +16,50 @@ use crate::serialize::*;
+    write_scalar(&self.ee, w)
   }
 
-  #[cfg(feature = "experimental")]
```
Keep experimental.

kayabaNerve commented on coins/monero/src/ringct/borromean.rs:
```diff
@@ -90,7 +82,6 @@ impl BorromeanRange {
     write_raw_vec(write_point, &self.bit_commitments, w)
   }
 
-  #[cfg(feature = "experimental")]
```
Keep experimental.

kayabaNerve commented on coins/monero/src/ringct/borromean.rs:
```diff
@@ -20,54 +16,50 @@ use crate::serialize::*;
+        &self.s1[i].recover_monero_slide_scalar(),
       );
-      transcript[i .. ((i + 1) * 32)].copy_from_slice(LV.compress().as_bytes());
+      transcript[i * 32 .. ((i + 1) * 32)].copy_from_slice(LV.compress().as_bytes());
```
I'd appreciate `(i * 32) ..`, if you don't mind.

Boog900 committed:
fix no-std

Boog900 committed:
Reduce the amount of rpc calls needed for `get_block_by_number`.
This function was causing me problems, every now and then a node would return a block with a different number than requested.

kayabaNerve commented:
I appreciate https://github.com/serai-dex/serai/pull/383/commits/46f43703a9b6d826cb27c6aeeeb53053143398bc, yet if you want to make further RPC changes, I'd appreciate it on a distinct PR.

By the way, one of the biggest changes we can make for performance will be moving to `get_blocks.bin`.

Boog900 committed:
change `serialize_hashable` to give the POW hashing blob.

Monero calculates the POW hash and the block hash using *slightly* different blobs :/

Boog900 committed:
make ring_signatures public and add length check when verifying.

kayabaNerve reviewed (commented):
This was a partial review I did weeks ago, sorry for the delay. I figured I may as well comment it now since I'm still rather occupied, sorry again. Mainly minor commentary :)

kayabaNerve committed:
Misc improvements and bug fixes

kayabaNerve commented:
Will merge once CI passes :+1: Thanks for the effort :)
